modules/top-level/test.nix

```diff
@@ -8,8 +8,79 @@
 let
   cfg = config.test;
 
-  inherit (config) warnings;
-  assertions = builtins.concatMap (x: lib.optional (!x.assertion) x.message) config.assertions;
+  # Expectation submodule used for checking warnings and/or assertions
+  expectation = lib.types.submodule (
+    { config, ... }:
+    {
+      options = {
+        predicate = lib.mkOption {
+          type = with lib.types; functionTo bool;
+          description = ''
+            A predicate of that determines whether this expectation is met.
+
+            Type
+            ```
+            [String] -> Boolean
+            ```
+
+            Parameters
+            1. values - all warnings or matched assertions
+          '';
+        };
+        expect = lib.mkOption {
+          type = with lib.types; nullOr (enum (builtins.attrNames cfg.predefinedExpectationPredicates));
+          description = ''
+            If non-null, will be used together with `value` to define `predicate`.
+          '';
+          default = "any";
+        };
+        value = lib.mkOption {
+          type =
+            if config.expect == null then
+              lib.types.unspecified
+              // {
+                description = "defined by `expect`";
+              }
+            else
+              cfg.predefinedExpectationPredicates.${config.expect}.valueType;
```

Comment:
That's some advanced depenedent type checking!

Reply:
I'd like this to to be better documented, we could actually fold `cfg.predefinedExpectationPredicates` into a bullet-point list of type-descriptions, but I figured that'd needlessly bloat an initial PR.